core/src/main/java/org/hibernate/ogm/dialect/eventstate/impl/EventContextManager.java

```diff
@@ -72,7 +72,9 @@ void onEventFinished() {
 
 		SessionImplementor session = (SessionImplementor) states.get( SessionImplementor.class );
 
-		for ( Entry<Class<?>, Object> state : states.entrySet() ) {
+		// we need to make a copy of the map as an OperationCollector entry might be added during
```

Comment:
A different approach would be to register all the necessary elements in the `onEventBegin` call. We could add an `onBegin` method to `EventStateLifecycle` and remove the create method.

Might be a good opportunity to only register the lifecycle if required by the dialect.